Readme.md
# Connect Riak Sessions
[![Build Status](https://travis-ci.org/randysecrist/connect-riak-sessions.png)](https://travis-ci.org/randysecrist/connect-riak-sessions)
[![Gem Version](https://badge.fury.io/js/connect-riak-sessions.png)](http://badge.fury.io/js/connect-riak-sessions)

connect-riak-sessions is a Riak session store implemented as middleware for [connect](https://github.com/senchalabs/connect) & [expressjs](https://github.com/visionmedia/express) backed by [nodiak](https://github.com/nathanaschbacher/nodiak).  Any version of Riak will work.

For extra bang for your buck; see automatic key expiration using [bitcask](http://docs.basho.com/riak/latest/ops/advanced/backends/bitcask).

 connect-riak-sessions `>= 0.0.1` developed on connect `2.9.2`.

## Installation

	  $ npm install connect-riak-sessions

## Options
  
  - `client` An existing nodiak client object you normally get from `riak.getClient()`
  - `bucket` Riak bucket to use defaulting to `expressjs_sessions`
  - `scheme` Riak scheme ('http'|'https')
  - `host` Riak server hostname
  - `port` Riak server tcp port #
  - ...    Remaining options passed to the riak `getClient()` method.

## Usage

To use with Connect or a Test Framework:

    var connect = require('express-session'),
        RiakStore = require('connect-riak-sessions')(connect);

    connect().use(connect.session({
      secret: 'keyboard cat',
      store: new RiakStore(options)
    }));
 
To use with ExpressJS:

    var RiakStore = require('connect-riak-sessions')(express);

    // Configure Middleware
    app.use(express.session({
      store: new RiakSessionStore({
        bucket: 'app_session_bucket',
        scheme: 'http',
        host: 'localhost',
        port: 8098,
      }),
      secret: 'keyboard cat'
    }));
    
## API Docs

The [API Docs](https://github.com/randysecrist/connect-riak-sessions/blob/master/API.md) are generated using [Dox](https://github.com/visionmedia/dox).

# License

  MIT


History.md

0.0.1 / 2013-10-19 
==================

  * initial work

0.0.2 / 2015-03-21
==================

  * use express-session
  * ensure tests pass


test.js

/**
 * Module dependencies.
 */

var assert = require('assert')
  , connect = require('connect')
  , RiakStore = require('./')(connect);

var store = new RiakStore({
  bucket: 'randy_test',
  scheme: 'http',
  host: 'localhost',
  port: 8098,
});

// #set()
store.set('123', { cookie: { maxAge: 2000 }, name: 'rks' }, function(err, ok){
  assert.ok(!err, '#set() got an error');
  assert.ok(ok, '#set() is not ok');

  // #get()
  store.get('123', function(err, data){
    assert.ok(!err, '#get() got an error');
    assert.deepEqual({ cookie: { maxAge: 2000 }, name: 'rks' }, data);

    // #destroy()
    store.destroy('123', function(){
      console.log('done');
      process.exit(0);
    });

    throw new Error('Error in fn');
  });
});

process.once('uncaughtException', function (err) {
  assert.ok(err.message === 'Error in fn', err.message);
});

lib/connect-riak.js
/*!
 * Connect - Riak
 * Copyright(c) 2013 Randy Secrist <randy.secrist@gmail.com>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var riak = require('nodiak')
  , debug = require('debug')('connect:riak');

/**
 * Return the `RiakStore` extending `connect`'s session Store.
 *
 * @param {object} connect
 * @return {Function}
 * @api public
 */

module.exports = function(connect){

  /**
   * Connect's Store.
   */

  var Store = connect.session.Store;

  /**
   * Initialize RiakStore with the given `options`.
   *
   * @param {Object} options
   * @api public
   */

  function RiakStore(options) {
    var self = this;

    options = options || {};
    Store.call(this, options);
    this.bucket = null == options.bucket
      ? 'expressjs_sessions'
      : options.bucket;

    debug('USING %s://%s:%s', options.scheme, options.host, options.port)
    this.client = options.client || new riak.getClient(options.scheme, options.host, options.port);
  };

  /**
   * Inherit from `Store`.
   */

  RiakStore.prototype.__proto__ = Store.prototype;

  /**
   * Attempt to fetch session by the given `sid`.
   *
   * @param {String} sid
   * @param {Function} fn
   * @api public
   */

  RiakStore.prototype.get = function(sid, fn){
    debug('GET "%s"', sid);
    var my_obj = this.client.bucket(this.bucket).objects.new(sid);
    my_obj.fetch(function(err, obj) {
      debug('GOT %s', obj.data);
      try {
        result = JSON.parse(obj.data); 
      }
      catch (err) {
        result = null;
      }
      return fn(null, result);
    });
  };

  /**
   * Commit the given `sess` object associated with the given `sid`.
   *
   * @param {String} sid
   * @param {Session} sess
   * @param {Function} fn
   * @api public
   */

  RiakStore.prototype.set = function(sid, sess, fn){
    try {
      var maxAge = sess.cookie.maxAge
        , ttl = this.ttl
        , sess = JSON.stringify(sess);

      ttl = ttl || ('number' == typeof maxAge
          ? maxAge / 1000 | 0
          : 86400); // oneDay in seconds

      debug('SET "%s" ttl:%s %s', sid, ttl, sess);
      var my_obj = this.client.bucket(this.bucket).objects.new(sid, sess);
      my_obj.save(function(err, obj) {
          err || debug('SET complete');
          fn && fn.apply(this, arguments);
      });
    }
    catch (err) {
      fn && fn(err);
    }
  };

  /**
   * Destroy the session associated with the given `sid`.
   *
   * @param {String} sid
   * @api public
   */

  RiakStore.prototype.destroy = function(sid, fn){
    debug('DESTROY "%s"', sid);
    var my_obj = this.client.bucket(this.bucket).objects.new(sid);
    my_obj.delete(function(err, obj) {
        debug('DESTROY complete');
    });
  };

  return RiakStore;
};


lib/connect-riak-sessions.js
/*!
 * Connect - Riak Sessions
 * Copyright(c) 2013 Randy Secrist <randy.secrist@gmail.com>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var riak = require('nodiak')
  , debug = require('debug')('connect:riak');

/**
 * Return the `RiakStore` extending `connect`'s session Store.
 *
 * @param {object} connect
 * @return {Function}
 * @api public
 */

module.exports = function(connect){

  /**
   * Connect's Store.
   */

  var Store = connect.session.Store;

  /**
   * Initialize RiakStore with the given `options`.
   *
   * @param {Object} options
   * @api public
   */

  function RiakStore(options) {
    var self = this;

    options = options || {};
    Store.call(this, options);
    this.bucket = null == options.bucket
      ? 'expressjs_sessions'
      : options.bucket;

    debug('USING %s://%s:%s', options.scheme, options.host, options.port)
    this.client = options.client || new riak.getClient(options.scheme, options.host, options.port);
  };

  /**
   * Inherit from `Store`.
   */

  RiakStore.prototype.__proto__ = Store.prototype;

  /**
   * Attempt to fetch session by the given `sid`.
   *
   * @param {String} sid
   * @param {Function} fn
   * @api public
   */

  RiakStore.prototype.get = function(sid, fn){
    debug('GET "%s"', sid);
    var my_obj = this.client.bucket(this.bucket).objects.new(sid);
    my_obj.fetch(function(err, obj) {
      debug('GOT %s', obj.data);
      try {
        result = JSON.parse(obj.data); 
      }
      catch (err) {
        result = null;
      }
      return fn(null, result);
    });
  };

  /**
   * Commit the given `sess` object associated with the given `sid`.
   *
   * @param {String} sid
   * @param {Session} sess
   * @param {Function} fn
   * @api public
   */

  RiakStore.prototype.set = function(sid, sess, fn){
    try {
      var maxAge = sess.cookie.maxAge
        , ttl = this.ttl
        , sess = JSON.stringify(sess);

      ttl = ttl || ('number' == typeof maxAge
          ? maxAge / 1000 | 0
          : 86400); // oneDay in seconds

      debug('SET "%s" ttl:%s %s', sid, ttl, sess);
      var my_obj = this.client.bucket(this.bucket).objects.new(sid, sess);
      my_obj.save(function(err, obj) {
          err || debug('SET complete');
          fn && fn.apply(this, arguments);
      });
    }
    catch (err) {
      fn && fn(err);
    }
  };

  /**
   * Destroy the session associated with the given `sid`.
   *
   * @param {String} sid
   * @api public
   */

  RiakStore.prototype.destroy = function(sid, fn){
    debug('DESTROY "%s"', sid);
    var my_obj = this.client.bucket(this.bucket).objects.new(sid);
    my_obj.delete(function(err, obj) {
        debug('DESTROY complete');
    });
  };

  return RiakStore;
};
